### PR TITLE
Implement Staking module

### PR DESCRIPTION
**Pull request by HoOngEe**
I tried to simulate staking logic previously is operating inside the host.
If it is merged into mold, the following tasks should be carried out.
1. Remove staking logic from the host.
2. Modify `TopLevelState` to store only CompactValidatorSet which contains only validators' public key and abstract voting powers. 
3. Remove parameters only used by `Staking`.

Followings are not implemented yet:
1. Executing `ReportDoubleVote` and banning malicious validators.
2. Creating transactions from higher level input (e.g. graphQL mutation request)

### CODE REVIEW

**Comment by HoOngEe** [comment]
Details will be modified after coordinator - module interface is fixed.

File: basic_module/staking/src/state.rs [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+    BTreeMap, BTreeSet, HashMap, HashSet,
+};
+use std::ops::Deref;
+
+type KEY = dyn AsRef<[u8]>;
+
+const STAKE_ACCOUNT_PREFIX: [u8; 1] = [0x1];
+const DELEGATION_PREFIX: [u8; 1] = [0x2];
+lazy_static! {
+    static ref METADATA_KEY: Vec<u8> = String::from("Metadata").into_bytes();
```
Comment (sgkim126) [inline basic_module/staking/src/state.rs:35]: Are these should be vectors?
If not, you don't need to use `lazy_static`.
```rust
static METADATA_KEY: &[u8; 8] = b"Metadata";
```

File: basic_module/staking/src/state.rs [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+impl NextValidators {
+    pub fn load() -> Self {
+        NextValidators(load_with_key(&*NEXT_VALIDATORS_KEY).unwrap_or_default())
+    }
+
+    pub fn save(self) {
+        write_with_key(&*NEXT_VALIDATORS_KEY, self.0)
+    }
+
+    pub fn elect() -> Self {
```
Comment (sgkim126) [inline basic_module/staking/src/state.rs:282]: FYI, currently, @majecty changes the way to store the validators' public keys.
https://github.com/CodeChain-io/foundry/pull/388

File: basic_module/staking/src/error.rs [hunk 3]
```diff
@@ -15,6 +15,7 @@
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
 pub use crate::runtime_error::Error as RuntimeError;
+pub use crate::syntax_error::Error as SyntaxError;
```
Comment (sgkim126) [inline basic_module/staking/src/error.rs:18]: Would you change the title to `Implement transaction and related types for the staking module`?

File: basic_module/staking/src/state.rs [hunk 4]
```diff
@@ -538,3 +559,14 @@ impl Banned {
         self.0.contains(public)
     }
 }
+
+pub fn get_stakes() -> HashMap<Public, u64> {
+    let stakeholders = Stakeholders::load();
+    let mut result = HashMap::new();
```
Comment (sgkim126) [inline basic_module/staking/src/state.rs:587]:
```rust
    stakeholders.iter().map(|stakeholder| {
        let account = StakeAccount::load(stakeholder);
        let delegation = Delegation::load(stakeholder);
        (*stakeholder, account.balance + delegation.sum())
    }).collect()
```

File: basic_module/staking/src/execute.rs [hunk 5]
```diff
@@ -276,3 +308,105 @@ pub fn change_params(
     metadata.save();
     Ok(outcome)
 }
+
+fn update_validators(validators: NextValidators) -> Result<TransactionExecutionOutcome, Error> {
+    let outcome: TransactionExecutionOutcome = Default::default();
```
Comment (sgkim126) [inline basic_module/staking/src/execute.rs:313]: I cannot find a reason to declare this variable here.

File: basic_module/staking/src/state.rs (hunk 1 above)
Comment (HoOngEe) [reply basic_module/staking/src/state.rs:35]: Ah, thank you. I'll apply your suggestion. They don't have to be vectors.

File: basic_module/staking/src/state.rs (hunk 4 above)
Comment (HoOngEe) [reply basic_module/staking/src/state.rs:587]: Thanks, I applied

File: basic_module/staking/src/error.rs (hunk 3 above)
Comment (HoOngEe) [reply basic_module/staking/src/error.rs:18]: Changed

File: basic_module/staking/src/execute.rs (hunk 5 above)
Comment (HoOngEe) [reply basic_module/staking/src/execute.rs:313]: The original intention was add events to the outcome as the function proceeds. However, I'm not sure about what events to emit so now empty. I'll move the default() to the return posititon.

File: basic_module/staking/src/state.rs (hunk 2 above)
Comment (HoOngEe) [reply basic_module/staking/src/state.rs:281]: I'll check it out and apply the change.

**Comment by HoOngEe** [comment]
@sgkim126 @majecty 
I checked all the modifications in the pr #388. Before applying the changes in staking module, I have two concerns.
1. `transaction_index` in a "block" could not be resolved from the module level. I naively thought that adding `transaction_index` when executing transactions based on the index in staking-transactions. However, it also produces another complexity. We're now using an interface function `prepare_transactions` which includes transaction executions to determine post-processing transactions(previously implicit) based on the state after execution. After the post-processing transactions are included, the index will be changed so the index when creating a block and the index after being delivered are different. Manually ignoring post-processing transactions when calculating the index is a possible way but unnatural.
2. Host-side validator ordering is important because it matches with the `bitset` in headers. In a modules system a host-side validator only stores a public key and an abstract voting power. To follow the new rule, host should additionally stores `nominated_at_block_number` and `nominated_at_transaction_index`. I don't think it's a general requirements for modules to determine those things as a tie-breaker.

**Comment by HoOngEe** [comment]
After a discussion @majecty 
I realized I misunderstood the boundary the second is no more a problem. The norm of host-side ordering is `public-key` so only staking module is entirely responsible for tie-breaking.
For the first one, I'll rename the variables and apply the change.

**Pull request merged**